pennylane/qnode.py

```diff
@@ -605,8 +605,25 @@ def __call__(self, *args, **kwargs):
             # construct the tape
             self.construct(args, kwargs)
 
-        # execute the tape
-        res = self.qtape.execute(device=self.device)
+        # Execute the tape.
+        # If the observable contains a Hamiltonian and the device does not
+        # support Hamiltonians, split tape into multiple tapes and execute
+        # sequentially. In future, this logic should be moved to the device
+        # to allow for more efficient batch execution.
+        supports_hamiltonian = self.device.supports_observable("Hamiltonian")
+        hamiltonian_in_obs = "Hamiltonian" in [obs.name for obs in self.qtape.observables]
+        if hamiltonian_in_obs and (self.device.shots is not None or not supports_hamiltonian):

```

Comment:
How come this isn't allowed in analytic mode?

Reply:
Hmm, I was a bit worried what happens if the device can handle analytic mode but not finite shots. But you are right, the device should decide here. Will change!